source/material_model/viscoelastic_plastic.cc

```diff
@@ -0,0 +1,796 @@
+            averaged_parameter = std::exp(averaged_parameter);
+            break;
+          }
+          case maximum_composition:
+          {
+            const unsigned int i = (unsigned int)(std::max_element( volume_fractions.begin(),
+                                                                    volume_fractions.end() )
+                                                  - volume_fractions.begin());
+            averaged_parameter = parameter_values[i];
+            break;
+          }
+          default:
+          {
+            AssertThrow( false, ExcNotImplemented() );
+            break;
+          }
+        }
+      return averaged_parameter;
+    }
+
+    template <int dim>
+    void
+    ViscoelasticPlastic<dim>::
+    evaluate(const MaterialModel::MaterialModelInputs<dim> &in,
+             MaterialModel::MaterialModelOutputs<dim> &out) const
+    {
+
+      // Store which components to exclude during volume fraction computation.
+      ComponentMask composition_mask(this->n_compositional_fields(),true);
+      // Elastic stress fields
+      for (unsigned int i = 0; i < SymmetricTensor<2,dim>::n_independent_components ; ++i)
+        composition_mask.set(i,false);
+
+      // Create the structure for the elastic force terms that are needed to compute the
+      // right-hand side of the Stokes system
+      MaterialModel::ElasticOutputs<dim>
+      *force_out = out.template get_additional_output<MaterialModel::ElasticOutputs<dim> >();
+
+      // The elastic time step (dte) is equal to the numerical time step if the time step number
+      // is greater than 0 and the parameter 'use_fixed_elastic_time_step' is set to false.
+      // On the first (0) time step the elastic time step is always equal to the value
+      // specified in 'fixed_elastic_time_step', which is also used in all subsequent time
+      // steps if 'use_fixed_elastic_time_step' is set to true.
+      const double dte = ( ( this->get_timestep_number() > 0 && use_fixed_elastic_time_step == false )
+                           ?
+                           this->get_timestep()
+                           :
+                           fixed_elastic_time_step);
+
+      for (unsigned int i=0; i < in.temperature.size(); ++i)
+        {
+
+          const double temperature = in.temperature[i];
+          const std::vector<double> composition = in.composition[i];
+          const double pressure = in.pressure[i];
+          const SymmetricTensor<2,dim> strain_rate = in.strain_rate[i];
+          const std::vector<double> volume_fractions = compute_volume_fractions(composition, composition_mask);
+
+          out.specific_heat[i] = average_value(volume_fractions, specific_heats, arithmetic);
+
+          // Arithmetic averaging of thermal conductivities
+          // This may not be strictly the most reasonable thing, but for most Earth materials we hope
+          // that they do not vary so much that it is a big problem.
```

Comment:
This comment also applies to the specific heat, density and thermal expansion, maybe put it a general statement above specific heat.